tmol/io/details/select_from_canonical.py

```diff
@@ -0,0 +1,430 @@
+    res_type_variants: Tensor[torch.int32][:, :],
+    found_disulfides64: Tensor[torch.int64][:, 3],
+    res_not_connected: Optional[Tensor[torch.bool][:, :, 2]] = None,
+) -> Tuple[
+    Tensor[torch.int64][:, :],
+    Tensor[torch.int64][:, :, :, 2],
+    Tensor[torch.int32][:, :, :, :, :],
+]:
+    pbt = packed_block_types
+    _annotate_packed_block_types_w_canonical_res_order(pbt)
+    _annotate_packed_block_types_w_dslf_conn_inds(pbt)
+    PoseStackBuilder._annotate_pbt_w_polymeric_down_up_bondsep_dist(pbt)
+    PoseStackBuilder._annotate_pbt_w_intraresidue_connection_atom_distances(pbt)
+
+    device = pbt.device
+    n_poses = chain_id.shape[0]
+    max_n_res = chain_id.shape[1]
+    max_n_conn = pbt.max_n_conn
+
+    # canonica_res_ordering_map dimensioned: [20aas x 3 termini types x 2 variant types]
+    # 3 termini types? 0-nterm, 1-mid, 2-cterm
+    canonical_res_ordering_map = pbt.canonical_res_ordering_map
+    res_types64 = res_types.to(torch.int64)
+    res_type_variants64 = res_type_variants.to(torch.int64)
+    is_real_res = res_types64 != -1
+
+    if res_not_connected is None:
+        res_not_connected = torch.zeros(
+            (n_poses, max_n_res, 2), dtype=torch.bool, device=device
+        )
+
+    # we are going to modify the chain_id tensor, perhaps to the detriment of the caller,
+    # so clone it.
+    # mark the chain id for any un-real residue as -1
+    # then the n-term residues will have a different chain id
+    # from their i-1 residues
+    chain_id = chain_id.clone()
+    chain_id[res_types64 == -1] = -1
+
+    # logic for deciding what chemical bonds are present between the polymeric
+    # residues and which residues should be represented as termini:
+    # - The first and last residues in a chain are not connected to the
+    #   previous / next residues
+    # - If res_not_connected[p, i, 0] is true, then this residue will not
+    #   be connected to i-1;
+    # - If res_not_connected[p, i, 1] is true, then this residue will not
+    #   be connected to i+1
+    # - Thus: for residue i and i+1 to be connected, all the following must hold:
+    #   - i cannot be the last residue in a chain
+    #   - (equivalently, i+1 cannot be the first residue in a chain)
+    #   - res_not_connected[p, i, 1] must be false
+    #   - res_not_connected[p, i+1, 0] must be false
+    # - If a residue is the first residue in a chain and res_not_connected[p, i, 0]
+    #   is true, then it will not be treated as an n-terminal residue, rather,
+    #   it will be treated as a "mid" residue (unless it is also a c-term residue)
+    #   and, in the case of amino acids, will have a regular H atom bonded to N
+    #   instead of the 1H, 2H, and 3H atoms. Its connection to the "previous" residue
+    #   will be incomplete (-1) instead of to a particular residue.
+    # - If a residue is the last residue in a chain and res_not_connected[p, i, 1]
+    #   is true, then it will not be treated as a c-terminal residue.
```

Comment:
Andrew, I don't like the "overloading" of res_not_connected and the interplay between chain and res_not_connected here.

Can we have an input Lx2 tensor to directly generate terminal variants? (e.g. if x[i,0]==true then i is an nterm variant)